go/host/enclave/service.go

```diff
@@ -200,24 +186,85 @@ func (e *Service) Unsubscribe(id rpc.ID) error {
 	return e.GetEnclaveClient().Unsubscribe(id)
 }
 
-// promoteNewActiveSequencer is a background goroutine that promotes a new active sequencer at startup or when the current one fails.
-// It will never give up, it just cycles through current enclaves until one can be successfully promoted.
-func (e *Service) promoteNewActiveSequencer() {
-	for e.activeSequencerID.Load() == _noActiveSequencer && e.running.Load() {
+// loadActiveSequencer is called at startup.
+// It will check to see if any of the enclaves think they are the Active Sequencer, and set the ID appropriately.
+// We don't need to worry about edge cases at this point, if something is wrong we will fix it in the
+// ensureActiveSequencer loop, but on the happy path this is much smoother
+func (e *Service) loadActiveSequencer() {
+	for _, guardian := range e.enclaveGuardians {
+		if guardian.state.IsEnclaveActiveSequencer() {
+			e.logger.Info("Found active sequencer enclave", log.EnclaveIDKey, guardian.GetEnclaveID())
+			e.activeSequencerID.Store(guardian.GetEnclaveID())
+			return
+		}
+	}
+}
+
+// ensureActiveSequencer is a background goroutine that attempts to ensure there is always exactly one
+// active sequencer enclave. It will promote a new active sequencer if the current one fails or is unavailable.
+func (e *Service) ensureActiveSequencer() {
+	for e.running.Load() {
+		activeSeqID := e.activeSequencerID.Load()
+		newSequencerRequired := true // this will be set false if we find the expected active sequencer is healthy
+
+		// do a pass through all guardians to see if they are in their expected state
 		for _, guardian := range e.enclaveGuardians {
-			enclID := guardian.GetEnclaveID()
-			e.logger.Info("Attempting to promote new sequencer.", log.EnclaveIDKey, enclID)
-			err := guardian.PromoteToActiveSequencer()
-			if err != nil {
-				e.logger.Info("Unable to promote new sequencer.", log.EnclaveIDKey, enclID, log.ErrKey, err)
-				continue
+			if guardian.state.IsEnclaveActiveSequencer() {
```

Comment:
I think this nested if thing could be more readable as a switch with the explicit condition for each case